### PR TITLE
Add NAS to Minitron pruning for parameter based auto-pruning

### PR DESCRIPTION
**Pull request by kevalmorabia97**
## What does this PR do?

**Type of change:** New feature 

- So far users we didnt have the NAS step from the Minitron paper so users had to manually find a pruned architecture which fits that params constraints and for that they have to tweak different combinations of width and depth pruning
- This PR adds a simplified version of the NAS from paper. We first find all candidate subnets that fit the user's params constraint and sort them by parameter count. 
  - [Paper] Then we pick top K candidates, do distillation for ~2B tokens and then select the one with best score (LM Loss / MMLU / other metric we care about). Note that the one among these top K with highest params is often not the best pruned model
  - [ModelOpt] Then we pick top K candidates, and select the one with best score (LM Loss / MMLU / other metric we care about). While doing KD gives better indication on which one to pick, skipping it makes the pruning much faster, much less compute intense, and finish everything in single prune API instead of first exporting top K models, doing KD and eval for all K models separately. We do print a Note in pruning step to let users know this so they can do KD if they want slightly better pruned model.
  - Further full KD is still needed as usual
- We also restrict the search space choices (e.g. `hidden_size` multiple of 256, `ffn_hidden_size` multiple of 512) to make the process efficient. Users can configure this if they want to.

## Usage

Pruning API is same as before:
```python
import modelopt.torch.prune as mtp
mtp.prune(
    model,
    mode="mcore_minitron",
    constraints=constraints,
    dummy_input=None,  # Not used
    config=config,
)
```

1. Manual Pruning (Existing):
```python
constraints = {"export_config": {"hidden_size: 3072", "ffn_hidden_size": 9216}}
config = {"forward_loop": forward_loop, "checkpoint": "/path/to/cache/pruning/scores.pth"}
mtp.prune(...)
```

2. NAS-based Auto Pruning (New):
```python
constraints = {"params": 6e9}. # prune to 6B params
config = {"forward_loop": forward_loop, "checkpoint": "/path/to/cache/pruning/scores.pth"}

# define the score_func to maximize (e.g MMLU, negative val loss, etc.)
from modelopt.torch.utils.plugins.megatron_mmlu import megatron_mmlu
def score_func(m):
    return megatron_mmlu(m, tokenizer, percentage=0.05)  # 5% sampled data for faster eval
config["score_func"] = score_func

# overwrite search space choices (showing defaults):
config["max_width_pruning"] = 0.4
config["max_depth_pruning"] = 0.2
config["hparams_to_skip"] = [] # can be used to disable pruning some hparams e.g. ["num_attention_heads"]
config["top_k"] = 10 # might be better to use 20 at the cost of longer time to prune

mtp.prune(...)
```

To configure search space (shows defaults):
```python
ss_config = mtp.mcore_minitron.get_mcore_minitron_config(
    hidden_size_divisor=256,
    ffn_hidden_size_divisor=512,
    mamba_head_dim_divisor=8,
    num_moe_experts_divisor=8,
    num_layers_divisor=2,
)
mtp.prune(model, mode=[("mcore_minitron", ss_config)], ....)
```

## Testing

**Qwen3-8B -> 6B (~2 hours on 8xA5000)**
```python
     0.4350 score -> {'num_layers': 34, 'hidden_size': 3328, 'ffn_hidden_size': 11264}
BEST 0.5705 score -> {'num_layers': 30, 'hidden_size': 3584, 'ffn_hidden_size': 11776}
     0.4051 score -> {'num_layers': 36, 'hidden_size': 3840, 'ffn_hidden_size': 8192}
     0.4593 score -> {'num_layers': 36, 'hidden_size': 3584, 'ffn_hidden_size': 9216}
     0.2737 score -> {'num_layers': 36, 'hidden_size': 3072, 'ffn_hidden_size': 11776}
     0.5556 score -> {'num_layers': 32, 'hidden_size': 3584, 'ffn_hidden_size': 10752}
     0.3198 score -> {'num_layers': 28, 'hidden_size': 4096, 'ffn_hidden_size': 10240}
     0.4119 score -> {'num_layers': 36, 'hidden_size': 4096, 'ffn_hidden_size': 7168}
     0.3808 score -> {'num_layers': 36, 'hidden_size': 3328, 'ffn_hidden_size': 10240}
     0.4783 score -> {'num_layers': 34, 'hidden_size': 3840, 'ffn_hidden_size': 8704}
```

**Nemotron-Nano-9B-v2 -> 7B (~2.5 hours on 8xA5000)**
```python
     0.2629 score -> {'num_layers': 54, 'hidden_size': 4352, 'mamba_num_heads':  88, 'mamba_head_dim': 72, 'ffn_hidden_size': 15360}
     0.2778 score -> {'num_layers': 50, 'hidden_size': 4480, 'mamba_num_heads': 128, 'mamba_head_dim': 56, 'ffn_hidden_size': 15680}
     0.5041 score -> {'num_layers': 56, 'hidden_size': 4096, 'mamba_num_heads':  96, 'mamba_head_dim': 80, 'ffn_hidden_size': 14336}
BEST 0.6043 score -> {'num_layers': 48, 'hidden_size': 4352, 'mamba_num_heads': 120, 'mamba_head_dim': 80, 'ffn_hidden_size': 13824}
     0.0772 score -> {'num_layers': 56, 'hidden_size': 4352, 'mamba_num_heads': 112, 'mamba_head_dim': 80, 'ffn_hidden_size': 10240}
     0.3550 score -> {'num_layers': 50, 'hidden_size': 4480, 'mamba_num_heads': 112, 'mamba_head_dim': 64, 'ffn_hidden_size': 15680}
     0.1016 score -> {'num_layers': 56, 'hidden_size': 4352, 'mamba_num_heads': 120, 'mamba_head_dim': 72, 'ffn_hidden_size': 10752}
     0.5461 score -> {'num_layers': 46, 'hidden_size': 4480, 'mamba_num_heads': 128, 'mamba_head_dim': 72, 'ffn_hidden_size': 14848}
     0.1992 score -> {'num_layers': 54, 'hidden_size': 4480, 'mamba_num_heads':  80, 'mamba_head_dim': 80, 'ffn_hidden_size': 14336}
     0.5881 score -> {'num_layers': 48, 'hidden_size': 4480, 'mamba_num_heads': 112, 'mamba_head_dim': 80, 'ffn_hidden_size': 13824}
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes 
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes 

## Additional Information

OMNIML-3043

## Release Notes

* **New Features**
  * Added NAS-based Auto Pruning for Minitron models as an alternative to manual pruning using parameter constraints
  * Introduced parameter counting capabilities for architecture search

* **Documentation**
  * Expanded pruning guides with detailed examples and workflows for both manual and automatic pruning approaches
  * Updated configuration documentation with granular divisor parameters

* **Improvements**
  * Enhanced parameter counting support for models with dynamic or mixture-of-experts modules

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by kevalmorabia97** [comment]
@coderabbitai review

File: modelopt/torch/prune/plugins/mcore_minitron.py [hunk 1]
```diff
@@ -264,36 +295,306 @@ def run_search(self) -> None:
...
+        # 1. Find available search space choices (across all PP ranks)
+        hp_choices = {}
+        for n, hp in named_hparams(self.model, configurable=True):
+            hp_name = n.split(".")[-1]
+            hp_choices[hp_name] = hp.choices
+        all_pp_search_spaces = [None] * get_pipeline_model_parallel_world_size()
+        torch.distributed.all_gather_object(
+            all_pp_search_spaces, hp_choices, group=get_pipeline_model_parallel_group()
+        )
+        hp_choices = {k: v for d in all_pp_search_spaces for k, v in d.items()}  # type: ignore[attr-defined]
```
Comment (realAsma) [inline modelopt/torch/prune/plugins/mcore_minitron.py]: nit:
Should we leverage ModelOpt utilities here?
```
pp_group = DistributedProcessGroup(get_pipeline_model_parallel_group())

hp_choices = DistributedProcessGroup.get_dist_syncd_obj(
    hp_choices, 
    pp_group, 
    op=lambda all_pp_search_spaces: {k: v for d in all_pp_search_spaces for k, v in d.items()}
)
```

File: modelopt/torch/utils/network.py [hunk 2]
```diff
@@ -112,6 +112,13 @@ def param_num(network: nn.Module, trainable_only: bool = False, unit=1e6) -> flo
     Returns:
         The number of parameters in the model in the given unit.
     """
+    from modelopt.torch.opt.dynamic import DynamicModule
+
+    if isinstance(network, DynamicModule):
```
Comment (realAsma) [inline modelopt/torch/utils/network.py]: this seems to be applicable only for pruning. Quantization have `QuantModule` base block - should we have something similar for pruning and do:
```
if isinstance(network, PruneModule):
```
or 
```
if hasattr(network, "some_special_pruning_attr")
```

File: modelopt/torch/prune/plugins/mcore_minitron.py (hunk 1 above)
Comment (kevalmorabia97) [reply modelopt/torch/prune/plugins/mcore_minitron.py]: Nice! Let me give this a try

File: modelopt/torch/utils/network.py (hunk 2 above)
Comment (kevalmorabia97) [reply modelopt/torch/utils/network.py]: Since quantization doesnt change param count, I dont see a reason to use this function. Besides, users can still use this on Quantized + Exported model since its not longer DynamicModule. Currently we will just raise an exception if Quantized DynamicModule but later if we need to support it, we can think of such a way. Wdyt?

File: modelopt/torch/prune/plugins/mcore_minitron.py (hunk 1 above)
Comment (kevalmorabia97) [reply modelopt/torch/prune/plugins/mcore_minitron.py]: Updated to use `DistributedProcessGroup`

**Comment by kevalmorabia97** [comment]
@coderabbitai review

File: modelopt/torch/utils/network.py (hunk 2 above)
Comment (realAsma) [reply modelopt/torch/utils/network.py]: I found this a bit a confusing. If this utility is only useful for pruning, should we move it to modelopt/torch/nas/utils.py?

I dont have strong opinion on this. I leave this to your discretion.

File: modelopt/torch/nas/plugins/megatron.py [hunk 3]
```diff
@@ -1045,27 +1021,30 @@ def modify(
...
-        hp = self.get_hparam("hidden_size")
-        choices = {int(make_divisible(c, hidden_size_divisor)) for c in hp.choices}  # type: ignore[arg-type]
-        hp.choices = list(set(hp.choices) & choices | {hp.original})
+        for hp_name, divisor in [
+            ("hidden_size", hidden_size_divisor),
+            ("num_layers", num_layers_divisor),
+        ]:
+            hp = self.get_hparam(hp_name)
+            choices = {int(make_divisible(c, divisor)) for c in hp.choices}  # type: ignore[arg-type]
+            hp.choices = list(set(hp.choices) & choices | {hp.original})
```
Comment (realAsma) [inline modelopt/torch/nas/plugins/megatron.py:1043]: quick question: I see the individual layers have a seperate param for `hidden_size` (a `TracedHparam`) - how are they linked? Where are we doing the tracing?

Comment (kevalmorabia97) [reply modelopt/torch/nas/plugins/megatron.py:1043]: https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/nas/plugins/megatron.py#L1020-L1040

I first extract the `hidden_size` hp from Embedding layers on rank 0. Then broadcast that hparam to all other ranks and for all subsequent layer DMRegistry.convert, I pass this hparam so its used instead of creating a new one for each layer

File: modelopt/torch/utils/network.py (hunk 2 above)
Comment (kevalmorabia97) [reply modelopt/torch/utils/network.py]: Will move to nas/utils

Comment (kevalmorabia97) [reply modelopt/torch/utils/network.py]: Actually, this function is only used in tests so will move to _test_utils
